Add build tools

meduzen opened the pull request:
Hi here!

The goal of this PR is to bring front-end tooling and validate that it works on maintainers machine.

Basically, it brings Webpack + ESLint + a set of config/tools for both CSS and JS. It doesn’t change any code architecture; this will come at a later stage.

Development instructions for the build tools in [README.md#development](https://github.com/DCMLab/reductive_analysis_app/blob/f15b9c80662324e628d0aa950ef42a8f87dc1a18/README.md#development)

yrammos commented:
@meduzen confirming that the output of the `prod` script is:

```
 ERROR  Failed to compile with 3 errors                                                      6:18:35 PM

 error  in ./src/js/vendor/wildwebmidi.js
…

 error  in ./src/js/vendor/wildwebmidi.js
...

 error  in ./src/js/vendor/wildwebmidi.js
...

webpack compiled with 3 errors in 44352 ms
...

```

pettter commented:
Hmm.. Even after getting node up to version 14 with nvm I'm not getting the same errors unfortunately..

```
[webpack-cli] HookWebpackError: Invalid string length
    at makeWebpackError (/home/ericson/jobb/repos/reductive_analysis_app/node_modules/webpack/lib/HookWebpackError.js:48:9)
    at /home/ericson/jobb/repos/reductive_analysis_app/node_modules/webpack/lib/Compilation.js:2510:12
    at eval (eval at create (/home/ericson/jobb/repos/reductive_analysis_app/node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:57:1)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
-- inner error --
RangeError: Invalid string length
    at Array.join (<anonymous>)
    at module.exports (/home/ericson/jobb/repos/reductive_analysis_app/node_modules/eslint/lib/cli-engine/formatters/codeframe.js:132:12)
    at Object.format (/home/ericson/jobb/repos/reductive_analysis_app/node_modules/eslint/lib/eslint/eslint.js:648:24)
    at formatResults (/home/ericson/jobb/repos/reductive_analysis_app/node_modules/eslint-webpack-plugin/dist/linter.js:194:49)
    at report (/home/ericson/jobb/repos/reductive_analysis_app/node_modules/eslint-webpack-plugin/dist/linter.js:119:9)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async processResults (/home/ericson/jobb/repos/reductive_analysis_app/node_modules/eslint-webpack-plugin/dist/index.js:149:13)
caused by plugins in Compilation.hooks.processAssets
RangeError: Invalid string length
    at Array.join (<anonymous>)
    at module.exports (/home/ericson/jobb/repos/reductive_analysis_app/node_modules/eslint/lib/cli-engine/formatters/codeframe.js:132:12)
    at Object.format (/home/ericson/jobb/repos/reductive_analysis_app/node_modules/eslint/lib/eslint/eslint.js:648:24)
    at formatResults (/home/ericson/jobb/repos/reductive_analysis_app/node_modules/eslint-webpack-plugin/dist/linter.js:194:49)
    at report (/home/ericson/jobb/repos/reductive_analysis_app/node_modules/eslint-webpack-plugin/dist/linter.js:119:9)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5)
    at async processResults (/home/ericson/jobb/repos/reductive_analysis_app/node_modules/eslint-webpack-plugin/dist/index.js:149:13)
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! @ production: `node_modules/webpack/bin/webpack.js --node-env=production --config=webpack.config.js`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the @ production script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

meduzen commented:
@pettter I think Webpack is trying to compile/optimize the Verovio library – which is heavy and may doom your machine – instead of simply copying it like instructed in the config file.

I moved libraries (Verovio and jQuery) directly to the `/public/` folder to solve that.

Now the compilation time is around 20 sec. (against 150 sec. before) on my machine. Can you retry?

I’ve also traded the wildwebmidi player for another one more suitable with Webpack bundling ([midi-player-js](https://github.com/grimmdude/MidiPlayerJS) + [soundfont-player](https://github.com/danigb/soundfont-player)). It works when we click “play”, but has no GUI for now. I’ll include it in a later commit or create a custom one at a later stage of the development.

Can you run again `npm i` and `npm run prod`?

If it passes, I’ll solve the last “bugs” before marking the PR reviewable and continuing on another branch.

yrammos commented:
Passing here:

` DONE  Compiled successfully in 19471ms`

pettter commented:
Interesting! I got a first ` DONE  Compiled successfully in 2963ms `, but then immediately after `Failed to compile with 1 error `, with a bunch of errors of this form:

`ERROR in error: Parsing error: 'import' and 'export' may appear only with 'sourceType: module' at src/js/app.js:1:1:`

yrammos commented:
Is this actually an eslint error?

meduzen commented:
It looks like an ESLint error indeed. It’s strange because the ESLint config already has `sourceType: module`. 🤔

@pettter I’d like to be sure the issue is not related to a previous Node version you used. Can you try this?

- Clear cache and reinstall: `rm -rf node_modules && rm package-lock.json && npm cache clean --force && npm install` (good time for a coffee).
- Compile stuff: `npm run prod`.

meduzen commented:
I’ve just pushed a commit to force the ECMA Version to latest. Shouldn’t be needed, but who knows. It also updates the lock file, but re-running `npm install` shouldn’t be needed.

pettter commented:
Will get that running asap after the meeting and let you know the result when we talk at two.

pettter commented:
Still getting the same errors with node 14, so I'm guessing something is reaching for system tools rather than the tools provided by the build system at some point.

pettter commented:
OK found it, it was an .eslintrc.js that was left over in the directory for some reason. My bad.

pettter commented:
```
 DONE  Compiled successfully in 8393ms                                                                                                                                           12:59:13 PM

hidden assets 0 bytes 1 asset
asset app.css?id=3951552f3b57c139b65e 1 bytes [emitted] [immutable] (name: app)
webpack compiled successfully in 2487 ms

asset app.js?id=ab28cc22bff294561786 302 KiB [emitted] [immutable] [minimized] [big] (name: app) 2 related assets
webpack compiled successfully in 8393 ms
```

meduzen commented:
Still left in this PR:

- don’t serve soundfont instrument from another server;
- bring back the player features: pause, stop, go to specific moment;
- fix trees;
- fix relations widths.

meduzen commented:
I think it’s now _on-par_ in terms of features.

- The player has been replaced by a custom one.
- I’ll make the player look better when working on implementing a design.
- Also: added UUIDs to layers and views, it’d be helping when going on with the UI. (c01df06)

yrammos commented:
@meduzen thanks for this. I have a couple of issues:

The rotating app loader ("spinner") never stops rotating, forcing me to disable it via the browser inspector.

After doing so, the UI appears broken.

I've run the following:

`rm -rf ./node_modules`
`npm i`
`npm run prod` (have also tried with `npm run dev` to be sure)
Cleared browser cache / force-reloaded all assets (though it shouldn't be necessary with webpack).

![Screen Shot 2021-07-25 at 09 58 42](https://user-images.githubusercontent.com/1519337/126890635-fb47bbd6-39c3-4bce-b309-312b89632e92.png)

yrammos commented:
Could it be that I'm using npm 7.20.1 — a bleeding-edge version?

yrammos commented on public/index.html:
```diff
@@ -0,0 +1,150 @@
+    <link type="text/css" rel="stylesheet" href="css/vendor/select2.min.css">
+    <link type="text/css" rel="stylesheet" href="css/vendor/style.css">
+    <link type="text/css" rel="stylesheet" href="css/app.css">
+    <script type="text/javascript" src="/js/vendor/jquery-3.1.1.min.js" defer></script>
+    <script type="text/javascript" src="/js/vendor/verovio-toolkit.js" defer></script>
+    <script type="text/javascript" src="/js/app.js" defer></script>
```
The first slash in front of the script paths should be removed, thus:

```js
<script type="text/javascript" src="js/vendor/jquery-3.1.1.min.js" defer></script>
<script type="text/javascript" src="js/vendor/verovio-toolkit.js" defer></script>
<script type="text/javascript" src="js/app.js" defer></script>
```

The scripts may fail to load otherwise.

meduzen replied:
That’s fixed. From what I see it only happened when you directly opened `index.html` in `file:///` or using `npm run start` (which is already available before of this PR).

- I remove the leading `/`.
- I’ve also updated the BrowserSync config for `npm run start` users.

meduzen commented:
Merge conflicts solved.

yrammos commented:
@meduzen I'm facing difficulties installing the npm modules.

```
❯ npm -v
7.20.1
❯ node -v
v14.16.1
```

I've cleared caches, lock file, modules etc. to no avail.

```
…
npm ERR! code 1
npm ERR! path /Users/rammos/Dev/reductive_analysis_app/node_modules/node-sass
npm ERR! command failed
npm ERR! command sh -c node scripts/install.js
npm ERR! internal/modules/cjs/loader.js:883
npm ERR!   throw err;
npm ERR!   ^
npm ERR!
npm ERR! Error: Cannot find module './pubsuffix-psl'
…
```

meduzen commented:
Mmmmh, indeed. I think I had downgraded my NPM version from 7 to 6 (because of work).

`jest-puppeteer` [seems to be the culprit](https://github.com/smooth-code/jest-puppeteer/issues/394). Updating it from version 4 to version 5 does the trick. As far as I can see, none of the [two breaking changes](https://github.com/smooth-code/jest-puppeteer/releases/tag/v5.0.0) needs additional changes.

Now noticing that none of the tests are passing. Gonna have a look during the evening.

meduzen commented:
> Now noticing that none of the tests are passing. Gonna have a look during the evening.

The tests don’t pass because:
- there’s no more globals, and the tests rely on them;
- querying elements by ID is broken for some elements because [they now may have a unique ID](https://github.com/DCMLab/reductive_analysis_app/blob/c01df067bf691bfdab81b4e02a84c81c49568168/src/js/app.js#L1).

Can we move forward without tests and fix them at a later stage?

pettter commented:
I'm not sure I understand why you introduced the new ID functionality in that commit? 

I'd like it if the tests worked (at least some of them) but I'd rather the functionality gets improved if doing so would take a lot of time.

meduzen commented:
> I'm not sure I understand why you introduced the new ID functionality in that commit?

Goo question. I think there was a bug without it. Gonna check later today or tomorrow.
 
> I'd like it if the tests worked (at least some of them) but I'd rather the functionality gets improved if doing so would take a lot of time.

It’ll probably take me a lot of time (my skills in end-to-end tests are currently limited to very basic stuff using [Cypress](https://www.cypress.io)) that I’d better spend on the UI in the coming weeks. At some point I’ll jump on that topic (later this year or next year) and practicing on this app and others will surely be a thing I’ll do.

meduzen commented:
> I'm not sure I understand why you introduced the new ID functionality in that commit?

@pettter I reverted it and couldn’t reproduce any bug, so I guess it’s fine to not have it for now.

I adapted the undo/redo feature (works great 👍 ) to the PR structure and fix various UI bugs that were introduced in my previous commits.

yrammos commented:
@meduzen I am still unable to build with the following toolchain:

```
❯ npm -v
7.20.1
❯ node -v
v14.16.1
```

Starting off with a clean slate:

```
❯ rm -rf node_modules && rm package-lock.json && npm cache clean --force && npm install
```

A number of modules are deprecated, by the way:

```
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
npm WARN deprecated debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
npm WARN deprecated sane@4.1.0: some dependency vulnerabilities fixed, support for node < 10 dropped, and newer ECMAScript syntax/features added
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated debug@4.1.1: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
```

Building fails as follows:

```
❯ npm run prod

> prod
> npm run production


> production
> node_modules/webpack/bin/webpack.js --node-env=production --config=webpack.config.js

[webpack-cli] Failed to load '/Users/rammos/Dev/reductive_analysis_app/webpack.config.js' config
[webpack-cli] TypeError: Cannot read property 'ES_LINT_AUTOFIX' of undefined
    at Object.<anonymous> (/Users/rammos/Dev/reductive_analysis_app/webpack.config.js:39:12)
    at Module._compile (/Users/rammos/Dev/reductive_analysis_app/node_modules/v8-compile-cache/v8-compile-cache.js:192:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (/Users/rammos/Dev/reductive_analysis_app/node_modules/v8-compile-cache/v8-compile-cache.js:159:20)
    at WebpackCLI.tryRequireThenImport (/Users/rammos/Dev/reductive_analysis_app/node_modules/webpack-cli/lib/webpack-cli.js:32:22)
    at loadConfig (/Users/rammos/Dev/reductive_analysis_app/node_modules/webpack-cli/lib/webpack-cli.js:1536:38)
    at /Users/rammos/Dev/reductive_analysis_app/node_modules/webpack-cli/lib/webpack-cli.js:1589:42
```

meduzen commented:
It seems you removed your local `.env` file (following a switch of branch, maybe), which is needed.

No worries about the deprecated libraries: they are only used by tooling (Babel, Jest, SASS…) and have no impact. Warning about security/deprecation are only important for apps being back-end apps running on Node, not for build tools that are not exposed in any way.

yrammos commented:
Duh, yes — thanks. Will be using the app today and report back.

yrammos commented:
@mehdi am error occurs when bringing relations to front (<key>z</key> shortcut with the  cursor over visually overlapping relations). Should be a simple fix. 

`ui.js:494`

meduzen commented:
Nice catch. 👍  Also learned about [`document.elementFromPoint`](https://developer.mozilla.org/en-US/docs/Web/API/Document/elementFromPoint) in the same time, so thanks!

yrammos commented:
@meduzen I ran into one small wrinkle and a rather more consequential bug:

1. Dragging the gray relations panel around in its *minimized* form (i.e. as a small square with a circle in the middle) automatically expands it at the end of the gesture. I realize you're planning to rework the UI, so this is a low-priority issue.

2. I believe that the drag-select functionality ("lasso" tool) is broken.

meduzen commented:
Okay, took me 3 hours of non-sense to find out that the last version of DragSelect is causing the issue with the lasso. 🥵

FYI I already started the new UI and have 2 branches ready to be PR’ed. I’m currently working on the filters, then I’ll move to the relations.

yrammos commented:
@meduzen I sympathize with your predicament… The vagaries of web-dev!

I hate to say that the lasso stills seems broken on my end. It works only as long as no note (or relation) has been previously (de)selected in the current session.

To reproduce:

- Open an analyzed score.
- Select a relation or a note.
- (Optional:) Deselect the relation or note.
- Try the lasso. On my end, at least, it's no longer working.

meduzen commented:
Forgot to mention: you have to run `npm install` (then compile) so that the _right_ version of DragSelect gets pulled.

I tried on the Chopin one (`mei-chopin.mei`) from the repository, and it works (with and without the optional step).

yrammos commented:
Ah, yes — I assumed `npm start` repacks the app before launching the server. It works now, sorry for the false alarm!

yrammos commented:
"Jump to next/previous bookmark" seems broken? (<kbd>{</kbd> and <kbd>}</kbd> with at least two bookmarked notes)

meduzen commented:
It’s working for me, but as I have an Azerty keyboard, there’s no way I can use `{` and `}` as keyboard shortcuts. When changing the shortcut keys, it works, and the menu buttons also work.

yrammos commented:
@meduzen thanks — it turns out that this feature is broken on my Safari Technology Preview but works well on Firefox (and presumably Chrome, as well). 

There is no regression in your PR (the bug is manifest in the `master` branch, too), so I'm leaving it up to you to determine the best PR in which to address it.

yrammos commented:
@meduzen May I ask whether the "naturalize note" feature works for you? To test:

- Open a score.
- Create a new layer.
- In the new layer, select an altered note (i.e. a note prefixed with an accidental # or ♭ or ♮).
- Press on "naturalize note" (<kbd>z</kbd>).

If the feature works, the accidental should be removed.

meduzen commented:
Indeed it wasn’t working. Now fixed. Also noticed it’s capital <kbd>Z</kbd>.

meduzen commented:
@pettter Following what we discussed a bit through another channel, can you create a `ui` branch based on `master`, so that I can change this PR target?

Optionally, I don’t know if it’s possible (or even if it’s a good idea 🤔) to give me access as “maintainer” of this `ui` branch, at least for 1 week so that I can bring the first chunks of work quicker.

pettter commented:
Done for the branch, I'll add you as a maintainer on the repo now (thought I already had, honestly).

meduzen commented:
Pushed a bunch of commits. Requires `npm install`.

- Added ability to expose stuff globally (variables or dependencies)
- Tests are now passing, but for some reasons, sometimes running the test is green 14/14, sometimes 13/14 (it says it fails loading the MEI score, but it doesn’t).
- Added production build to the repository.

yrammos commented:
@meduzen a number of tests are failing on my end. I ran:

`$ gh pr checkout 130`
`$ npm install`
`$ cp .env.example .env`
`$ npm run prod`
`…`
`webpack compiled successfully in 17096 ms`
`$ npm run test`
`…`
```
 FAIL  ./index.test.js (17.982 s)
  reductive_analysis_test_suite
    ✓ should run a rudimentary test on static HTML to confirm Jest works (9 ms)
    ✓ should parse conf.js without throwing an exception (5 ms)
    ✓ should set up all buttons with expected element IDs and attributes (1 ms)
    ✓ should load the example MEI (138 ms)
    ✓ should have loaded view-specific buttons
    ✕ should produce a directed <graph> within <mei> (5 ms)
    ✕ should produce a convincing <mei> object (using Jest snapshots) (17 ms)
    ✕ should produce a convincing SVG (using Jest snapshots) (5 ms)
    ✕ should ensure that (the very last) note IDs of the MEI and the SVG match (19 ms)
    ✕ should toggle a note, ensuring that the relevant array is updated and the note styled accordingly (10003 ms)
    ✓ should extra-toggle a note, ensuring that the relevant array is updated and the note styled accordingly (289 ms)
    ✓ should toggle a new relation between structurally unequal notes (264 ms)
    ✓ should reduce the relation between structurally unequal notes, hiding the less important one (161 ms)
    ✓ should unreduce the relation, showing it again (114 ms)
```

Running the tests again reproduces the same errors.

pettter commented:
Hm.. For me it works, mostly. I wonder if upping the times waiting for the loading of UI and score could help?

yrammos commented:
`should load the example MEI` passes on my end. The problem seems to be that `$(window.mei).children()` is empty… At least this is what Jest claims. Will investigate a little more.

pettter commented:
OK, it seems that turning logging on and off (`verbose`at line 17) definitely has an effect at which tests pass, possibly by affecting some race conditions through `console.log` influencing serialization or something. At the moment I can reliably get either the first or the fourth test to fail, but not both to pass, for some reason.

I don't know what's going on.

pettter commented:
There's definitely some race condition shenanigans.

Running `jest --runInBand` fails the fourth test.

Running `jest --maxWorkers=2` fails the first test.

yrammos commented:
Might it be webpack?

If I `npm run development` rather than `npm run prod`, then run `npm run test`, all tests pass.

yrammos commented:
(They pass with `verbose = true`. If `false`, two tests still fail (namely the note (extra)-toggling tests.)

yrammos commented:
I suspect one or more `await` is missing and thus the corresponding tests pass prematurely, on occasion.

pettter commented:
This branch is now more or less ready for merge into `ui`, I think, since the tests are working (with some caveats).
